File: ios/OffgridMobile/PDFExtractor/PDFExtractorModule.swift
import Foundation
import PDFKit

@objc(PDFExtractorModule)
class PDFExtractorModule: NSObject {

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func extractText(_ filePath: String, maxChars: Double, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .userInitiated).async {
      guard let url = URL(string: filePath) ?? URL(fileURLWithPath: filePath) as URL?,
            let document = PDFDocument(url: url) else {
        reject("PDF_ERROR", "Could not open PDF file", nil)
        return
      }

      let limit = Int(maxChars)
      var fullText = ""
      for pageIndex in 0..<document.pageCount {
        if let page = document.page(at: pageIndex), let pageText = page.string {
          fullText += pageText
          if pageIndex < document.pageCount - 1 {
            fullText += "\n\n"
          }
        }

        if fullText.count >= limit {
          fullText = String(fullText.prefix(limit))
          fullText += "\n\n... [Extracted \(pageIndex + 1) of \(document.pageCount) pages]"
          break
        }
      }

      resolve(fullText)
    }
  }
}

File: android/app/src/main/java/ai/offgridmobile/pdf/PDFExtractorModule.kt
package ai.offgridmobile.pdf

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import io.legere.pdfiumandroid.PdfDocument
import io.legere.pdfiumandroid.PdfiumCore
import android.os.ParcelFileDescriptor
import java.io.File

class PDFExtractorModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "PDFExtractorModule"

    private fun extractPageText(doc: PdfDocument, pageIndex: Int, sb: StringBuilder) {
        val page = doc.openPage(pageIndex)
        val textPage = page.openTextPage()
        val charCount = textPage.textPageCountChars()
        if (charCount > 0) {
            val text = textPage.textPageGetText(0, charCount)
            if (text != null) sb.append(text).append("\n\n")
        }
        textPage.close()
        page.close()
    }

    @ReactMethod
    fun extractText(filePath: String, maxChars: Double, promise: Promise) {
        Thread {
            try {
                val file = File(filePath)
                if (!file.exists()) {
                    promise.reject("PDF_ERROR", "File not found: $filePath")
                    return@Thread
                }

                val limit = maxChars.toInt()
                val core = PdfiumCore(reactApplicationContext)
                val fd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
                val doc = core.newDocument(fd)
                val pageCount = doc.getPageCount()
                val sb = StringBuilder()

                for (i in 0 until pageCount) {
                    extractPageText(doc, i, sb)

                    if (sb.length >= limit) {
                        sb.setLength(limit)
                        sb.append("\n\n... [Extracted ${i + 1} of $pageCount pages]")
                        break
                    }
                }

                doc.close()
                fd.close()
                promise.resolve(sb.toString())
            } catch (e: Exception) {
                promise.reject("PDF_ERROR", "Failed to extract text: ${e.message}", e)
            }
        }.start()
    }
}
